File: BOJ_7576_토마토/7576_허자용.py
import collections

green = 0
q = collections.deque()
box = []
dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]


def isOutOfBox(r, c):
    if r == -1 or r == N or c == -1 or c == M:
        return True
    else:
        return False


def ripe(r, c, date):
    for i in range(4):
        if isOutOfBox(r + dr[i], c + dc[i]):
            continue
        elif box[r + dr[i]][c + dc[i]] == 0:
            box[r + dr[i]][c + dc[i]] = 1
            global green
            green -= 1
            q.append(r + dr[i])
            q.append(c + dc[i])
            q.append(date)


[M, N] = list(map(int, input().split()))

for i in range(N):
    templist = list(map(int,input().split()))
    box.append(templist)
    for j in range(M):
        if not box[i][j]:
            green += 1
        if box[i][j] == 1:
            q.append(i)
            q.append(j)
            q.append(0)

if not green:
    print(0)

else:
    while(q):
        temp_r = q.popleft()
        temp_c = q.popleft()
        temp_date = q.popleft()
        ripe(temp_r, temp_c, temp_date + 1)

    if green:
        print(-1)

    else:
        print(temp_date)

File: BOJ_7576_토마토/7576_허자용.cpp
#include <iostream>
#include <queue>

using namespace std;

int M, N;
int box[1000][1000];
int green = 0;
int temp, temp_r, temp_c, temp_date;
queue<int> q;
int dr[4] = {1, -1, 0, 0};
int dc[4] = {0, 0, 1, -1};

bool isOutOfBox(int r, int c) {
    if(r == -1 || r == N || c == -1 || c == M) return true;
    else return false;
}

void ripe(int r, int c, int date) {
    for(int i = 0; i < 4; i++) {
        if(isOutOfBox(r + dr[i], c + dc[i])) continue;
        else if(box[r + dr[i]][c + dc[i]] == 0) {
            box[r + dr[i]][c + dc[i]] = 1;
            green--;
            q.push(r + dr[i]);
            q.push(c + dc[i]);
            q.push(date);
        }
    }
}

int main() {
    cin.sync_with_stdio(false);
    cin >> M >> N;
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < M; j++) {
            cin >> temp;
            box[i][j] = temp;
            if(!temp) green++;
            if(temp == 1) {
                q.push(i);
                q.push(j);
                q.push(0);
            }
        }
    }
    
    if(!green) {
        cout << 0 << endl;
        return 0;
    }

    while(!q.empty()) {
        temp_r = q.front();
        q.pop();
        temp_c = q.front();
        q.pop();
        temp_date = q.front();
        q.pop();
        ripe(temp_r, temp_c, temp_date + 1);
    }
    
    if(green) {
        cout << -1 << endl;
        return 0;
    }

    cout << temp_date << endl;
    return 0;
  }